Fix recursive comparison operator for C++20

bobbymcr opened the pull request:
Change `(x != y)` to `!(x == y)` to resolve ambiguity leading to recursive comparison call.

Fixes #2150

tencent-adm commented:
[![CLA assistant check](https://cla.opensource.tencent.com/pull/badge/not_signed)](https://cla.opensource.tencent.com/Tencent/rapidjson?pullRequest=2197) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/rapidjson?pullRequest=2197) before we can accept your contribution.<br/><hr/>**Brian Rogers** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/rapidjson?pullRequest=2197) it.</sub>

bobbymcr committed:
Fix recursive comparison operator for C++20

Change `(x != y)` to `!(x == y)` to resolve ambiguity leading to recursive
comparison call.

Fixes #2150
